common/amount.go

```diff
@@ -0,0 +1,79 @@
+package common
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Amount can be used to convert from fixed point to integer.
+// Amount assumes the smallest divisible unit is 1e-8
+type Amount int64
+
+func (a Amount) AsInt64() int64 {
+	return int64(a)
+}
+
+func (a Amount) MarshalJSON() ([]byte, error) {
+	return []byte(AmountToString(int64(a))), nil
+}
+
+func (a *Amount) UnmarshalJSON(data []byte) error {
+	v, err := StringToAmount(string(data))
+	if err != nil {
+		return err
+	}
+	*a = Amount(v)
+	return nil
+}
+
+func FloatToAmount(f float64) int64 {
+	return int64(f * 1e8)
+}
+
+// StringToAmount takes a number in decimal form, and converts it to
+// an integer * 1e8. Does not accept negative values
+func StringToAmount(amt string) (int64, error) {
+	valid := regexp.MustCompile(`^([0-9]+)?(\.[0-9]+)?$`)
+	if !valid.MatchString(amt) {
+		return 0, fmt.Errorf("improper string")
+	}
+
+	var total int64 = 0
+
+	dot := regexp.MustCompile(`\.`)
+	pieces := dot.Split(amt, 2)
+	whole, _ := strconv.Atoi(pieces[0])
+	total += int64(whole) * 1e8
```

Comment:
This regex allows an empty string which produces an error that you're suppressing in line 48. It should be `^[0-9]+(\.[0-9]+)?$`. to force it to be "number", or "number.number". It also doesn't include edge cases like ".5" or "10.". 

Further, your conversion to int ignores all overflow errors which could lead to div by 0 errors or other oddities. It's also not safe for 32-bit systems since you're converting to golang's int.

Reply:
I'll make these changes and add unit tests for those.

Reply:
Wait, I have that regex above. This is just for the splitting, if there is no decimal point, then the split results in 1 sections.

`10.` is deemed invalid though.

I added better overflow checking